drivers/w1/w1_max32.c

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright (c) 2023-2024 Analog Devices, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/w1.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/clock_control/adi_max32_clock_control.h>
+
+#include <wrap_max32_owm.h>
+
+#define DT_DRV_COMPAT adi_max32_w1
+
+LOG_MODULE_REGISTER(w1_max32, CONFIG_W1_LOG_LEVEL);
+
+struct max32_w1_config {
+	struct w1_master_config w1_config;
+	mxc_owm_regs_t *regs;
+	const struct pinctrl_dev_config *pctrl;
+	const struct device *clock;
+	struct max32_perclk perclk;
+	int internal_pullup;
+	int external_pullup;
+	int long_line_mode;
+};
+
+struct max32_w1_data {
+	struct w1_master_data w1_data;
+	uint8_t reg_device_config;
+};
+
+static int api_reset_bus(const struct device *dev)
+{
+	int ret;
+	const struct max32_w1_config *const cfg = dev->config;
+	mxc_owm_regs_t *regs = cfg->regs;
+
+	/* 0 if no 1-wire devices responded during the presence pulse, 1 otherwise */
+	ret = MXC_OWM_Reset();
+	if (ret > 0) {
+		/* Check OW pin input state due to presence detect pin seems not work well */
+		if (regs->ctrl_stat & MXC_F_OWM_CTRL_STAT_OW_INPUT) {
+			ret = 1; /* At least 1 device on the line */
+		} else {
+			ret = 0; /* no device on the line */
+		}
+	}
+
+	return ret;
+}
+
+static int api_read_bit(const struct device *dev)
+{
+	int ret;
+
+	ret = MXC_OWM_ReadBit();
+	if (ret < 0) {
+		if (MXC_OWM_GetPresenceDetect() == 0) {
+			/* if no slave connected to the bus, read bits shall be logical ones */
+			ret = 1;
+		} else {
+			return -EIO;
+		}
```

Comment:
I'm trying to understand the behaviour we will have here without having access to the hardware:
Looking at the read_bit implementation of owm_reva the only time a negative value is returned in case of a short on the line. There are a few details I can't derive from the datasheet. Assuming that short was already present in the last reset, would the present flag then also always be set?

Reply:
Yes, if the line being short it return negative value, https://github.com/zephyrproject-rtos/hal_adi/blob/main/MAX/Libraries/PeriphDrivers/Source/OWM/owm_reva.c#L156
The present flag being clear in that case, https://github.com/zephyrproject-rtos/hal_adi/blob/main/MAX/Libraries/PeriphDrivers/Source/OWM/owm_me18.c#L92C25-L92C35